library/core/src/panicking.rs

```diff
@@ -92,6 +93,39 @@ pub fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
     unsafe { panic_impl(&pi) }
 }
 
+/// The underlying implementation of libcore's `panic!` macro when formatting is used.
+#[cold]
+#[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
+#[cfg_attr(feature = "panic_immediate_abort", inline)]
+#[track_caller]
+#[cfg(not(bootstrap))]
+pub const fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
+    if cfg!(feature = "panic_immediate_abort") {
+        super::intrinsics::abort()
+    }
+
+    // NOTE This function never crosses the FFI boundary; it's a Rust-to-Rust call
+    // that gets resolved to the `#[panic_handler]` function.
+    extern "Rust" {
+        #[lang = "panic_impl"]
+        #[rustc_const_unstable(feature = "const_panic_impl", issue = "none")]
+        fn panic_impl(pi: &PanicInfo<'_>) -> !;
+    }
+
+    #[cfg(not(bootstrap))]
+    // SAFETY: the intrinsic is always safe to call.
+    unsafe {
+        // If we're const-evaluating this panic, this call will abort evaluation and unwind.
+        // The code below is only reachable during runtime.
+        core::intrinsics::panic_ctfe_hook(fmt.as_str());
```

Comment:
Would it make sense to move this up all the way above even the `panic_immediate_abort` check? Then CTFE panics will behave correctly even with that flag.

`panic_ctfe_hook` compiles to a runtime NOP so this still makes it immediately abort.